modules/dnn/src/ie_ngraph.hpp

```diff
@@ -0,0 +1,144 @@
+
+#ifdef HAVE_DNN_NGRAPH
+
+#include <ngraph/ngraph.hpp>
+
+#endif  // HAVE_DNN_NGRAPH
+
+namespace cv { namespace dnn {
+
+#ifdef HAVE_DNN_NGRAPH
+
+class InfEngineNgraphNode;
+
+
+class InfEngineNgraphNet
+{
+public:
+    InfEngineNgraphNet();
+    InfEngineNgraphNet(InferenceEngine::CNNNetwork& net);
+
+    void addOutput(const std::string& name);
+
+    bool isInitialized();
+    void init(Target targetId);
+
+    void forward(const std::vector<Ptr<BackendWrapper> >& outBlobsWrappers, bool isAsync);
+
+    void initPlugin(InferenceEngine::CNNNetwork& net);
+    ngraph::ParameterVector setInputs(const std::vector<cv::Mat>& inputs, const std::vector<std::string>& names);
+
+    void setUnconnectedNodes(Ptr<InfEngineNgraphNode>& node);
+    void addBlobs(const std::vector<cv::Ptr<BackendWrapper> >& ptrs);
+
+    void createNet(Target targetId);
+    void setNodePtr(std::shared_ptr<ngraph::Node>* ptr);
+private:
+    void release();
+    int getNumComponents();
+    void dfs(std::shared_ptr<ngraph::Node>& node, std::vector<std::shared_ptr<ngraph::Node>>& comp,
+             std::unordered_map<std::string, bool>& used);
+
+    ngraph::ParameterVector inputs_vec;
+    std::shared_ptr<ngraph::Function> ngraph_function;
+    std::vector<std::vector<std::shared_ptr<ngraph::Node>>> components;
+    std::unordered_map<std::string, std::shared_ptr<ngraph::Node>* > all_nodes;
+
+    InferenceEngine::ExecutableNetwork netExec;
+    InferenceEngine::BlobMap allBlobs;
+    std::string device_name;
+    bool isInit = false;
+
+    struct NgraphReqWrapper
+    {
+        NgraphReqWrapper() : isReady(true) {}
+
+        void makePromises(const std::vector<Ptr<BackendWrapper> >& outs);
+
+        InferenceEngine::InferRequest req;
+        std::vector<cv::AsyncPromise> outProms;
+        std::vector<std::string> outsNames;
+        bool isReady;
+    };
+    std::vector<Ptr<NgraphReqWrapper> > infRequests;
+
+    InferenceEngine::CNNNetwork cnn;
+    bool hasNetOwner;
+    std::vector<std::string> requestedOutputs;
+    std::unordered_set<std::shared_ptr<ngraph::Node>> unconnectedNodes;
+};
+
+class InfEngineNgraphNode : public BackendNode
+{
+public:
+    InfEngineNgraphNode(std::shared_ptr<ngraph::Node>&& _node);
+    InfEngineNgraphNode(std::shared_ptr<ngraph::Node>& _node);
+
+    void setName(const std::string& name);
+
+    // Inference Engine network object that allows to obtain the outputs of this layer.
+    std::shared_ptr<ngraph::Node> node;
+    Ptr<InfEngineNgraphNet> net;
+};
+
+class NgraphBackendWrapper : public BackendWrapper
+{
+public:
+    NgraphBackendWrapper(int targetId, const Mat& m);
+    NgraphBackendWrapper(Ptr<BackendWrapper> wrapper);
+    ~NgraphBackendWrapper();
+
+    static Ptr<BackendWrapper> create(Ptr<BackendWrapper> wrapper);
+
+    virtual void copyToHost() CV_OVERRIDE;
+    virtual void setHostDirty() CV_OVERRIDE;
+
+    InferenceEngine::DataPtr dataPtr;
+    InferenceEngine::Blob::Ptr blob;
+    AsyncArray futureMat;
+};
```

Comment:
Can we merge `NgraphBackendWrapper` and Inference Engine's wrapper?

Reply:
I believe it is better to "replace" it completely instead (after removal of NN Builder API).

I prefer to keep them separate - it least to guarantee that we don't break NN Builder API for now.

Reply:
Sure, thank you!